### PR TITLE
fix: respect root boundaries and native event cancellation

### PR DESCRIPTION
**Pull request by domgan-oai**
## Summary

- Bound delegated capture/bubble queues to native root delivery, so external listeners between roots interleave before the next root and can stop it.
- Traverse observer-specific composed paths for shadow roots and distributed slots while preserving native event identity, retargeting, and logical portal ancestry.
- Separate framework propagation from native stop flags: an earlier same-root native `stopPropagation()` does not truncate the framework queue; native `stopImmediatePropagation()` does not truncate an already-running logical phase. Call both methods to stop both queues.
- Preserve root boundaries when a handler unmounts/reinserts a nested root during dispatch; restore temporary method descriptors, support same-Event redispatch, and expose native dialog lifecycle handlers on logical ancestors in JSX types.

- Recover ownership for portal hosts that appear after child-local updates, including shared targets and native listeners moving the host before bubbling.

- Unregister compiled-portal ownership during scope teardown, with a deterministic shared-target lifecycle cleanup regression.

## Why

Follow-up to #925, based on independently reproduced event-system interop defects. This does not claim to resolve the original application-level missing-submit report.

The former dispatcher stamped each native event as handled and walked the entire logical ancestry from the first listener. That crossed intervening native listeners, skipped shadow/slot handlers, and treated an external native stop flag as cancellation of the framework's own queue.

## Validation

- [x] Current-head CI: all 26 jobs passed on `c3c476eb0` ([run 33567573036](https://github.com/octanejs/octane/actions/runs/33567573036)); current-head Cursor Bugbot passed with no issues.
- [ ] Repository-wide `pnpm format:check` (scoped formatting used instead).
- [ ] Repository-wide `pnpm typecheck` (scoped source, fixture, browser, and JSX type checks used instead).
- [ ] All-workspace `pnpm test` (both complete core Vitest projects used locally; CI covers the workspace).
- [x] Scoped `format-files.mjs --check` and `typecheck-files.mjs` with the existing imported `slot-hooks.js` included for the SSR test loader.
- [x] `vitest run --project octane --project octane-prod --project radix --project shadcn --project vaul`: 15,896 tests passed across 937 files.
- [x] Trusted Chromium event-boundary, continuous-action, and Vaul snap-drag suites: 24 tests passed, including shared React 19.2.7 oracle cases.
- [x] Full Visx package typecheck; native lifecycle additions are normalized precisely and pinned to optional native handlers, with a negative synthetic-wrapper check.
- [x] New work-only benchmark helper/fixture passed strict isolated `tsrx-tsc`; no existing TypeScript project owns these benchmark files.
- [x] Compiled-portal cleanup regression: the prior head performed 256 leftover sibling reads after six owners unmounted; the final source performs zero with identical event/ref/DOM controls.
- [x] `pnpm sync` and `git diff --check`.
- [x] Red evidence: original conformance regressions failed in both compile modes; 8 of 15 initial Chromium cases failed; 8 unmount/reinsert cases reproduced duplicate inner handlers; the new capture-immediate case failed in both modes on the base. A dropped-handler mutation made 78 conformance assertions fail, then the candidate was restored. The first CI run exposed late portal ownership loss in Vaul, Radix Select, and shadcn; eight added dev/prod regressions fail on the initial PR head and pass on the corrected runtime. The existing binding/browser tests remain unchanged.

### Performance

Paired production Chromium runs against base `f6bea3704`, identical dependency graph, 30 samples per version, macOS arm64 / Node 26.4.0. Commands: `node benchmarks/event-delegation/work.mjs` and `node benchmarks/event-delegation/run.mjs octane-tsrx 30`.

| Metric | Base | Candidate |
| --- | ---: | ---: |
| 128-input burst mean | 6.07 ms | 6.69 ms |
| Median | 7.10 ms | 7.50 ms |
| p95 | 7.60 ms | 8.20 ms |
| Mean relative margin of error | 9.86% | 8.30% |
| Four fixture JS assets, raw | 197,772 B | 202,216 B |
| Same assets, gzip | 63,889 B | 65,273 B |

The timing difference is within observed variance; there is no speedup or no-regression claim. The definite bundle cost is +1,384 gzip bytes. Both deterministic gates pass: 512 hosts, 128 native captures/bubbles/framework captures, 128 correct controlled inputs/outputs, no invalid currentTargets, and 256 currentTarget definitions sharing one descriptor/getter/path. Ordinary single-root dispatch avoids foreign-root ownership lookups; only emulated bubbling interposes native immediate cancellation. Root-history allocations and cleanup are confined to root lifecycle changes; events retain numeric epochs, not root sets.

A separate exploratory production Chromium probe used 20 alternating paired samples after five warmups, 1,000 untrusted native events per sample, and no state updates. All eight placement, logical phase-order, native identity/currentTarget, and teardown controls passed. With 200 body portals, portal clicks measured 2.51 → 3.13 µs/event and input in an unrelated root before the ranges measured 2.56 → 4.49 µs/event. An earlier correction scanned every range; the final boundary walk removes those repeated DOM-order comparisons. It is not order-independent constant work: a root after the 200 ranges measured 2.59 → 8.80 µs/event (201 sibling reads/event), or 9.24 µs with 64 additional preceding siblings (265 reads/event). These exploratory comparisons include the broader PR changes, not only the late-ownership correction. No negative cache assumes the DOM stayed put.

The work gate also covers six compiled portal owners over three shared-target cycles: nine deliveries in each logical phase, six ref cleanups, no detached clicks or leftover nodes, and zero sibling/DOM-order work on subsequent ordinary inputs. The ordinary timing fixture does not include these intrinsic observers.

## Risk / follow-ups

- Intentional cancellation behavior change: consumers using only native `stopImmediatePropagation()` to stop logical propagation must also call `stopPropagation()`. Direct native prototype calls, `cancelBubble` assignment, and immutable own stop methods remain native-only.
- Closed-shadow portals physically inside their own outer root can still duplicate shared outer delivery because the outer observer cannot see the hidden portal boundary. A trusted-click probe reproduced the same closed-shadow duplication in React 19.2.7. No universal React parity is claimed for shadow/portal cross-products or emulated non-bubbling plugin timing.
- Discrete work now flushes at each delivered root bubble boundary. Explicit transitions, pending-Action isolation, controlled restoration, hydration, portals, error handling, and reentrancy are covered by the core and browser runs.
- Chromium was exercised locally; other browser engines were not measured. Root lifecycle changes before the first framework capture observer are not a claim of full React Fiber ownership equivalence.

## Provenance

- [x] An agent produced this diff (`agent-authored`)

---
<a href="https://app.blacksmith.sh/octanejs/codesmith/octane/pr/926"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1790891611&installation_model_id=432790&pr_number=926&repository=octanejs%2Foctane&return_to=https%3A%2F%2Fgithub.com%2Foctanejs%2Foctane%2Fpull%2F926&signature=c8dcf3c0e4161c5c2bfc0b13db56760725728f64770816df7cf8f503e685aa45"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---

> [!NOTE]
> **High Risk**
> Rewrites core delegated event dispatch, nested-root ordering, portal ownership, and stop-propagation semantics—behavior changes that affect every interactive app and require callers who relied on native `stopImmediatePropagation()` alone to also call `stopPropagation()`.
> 
> **Overview**
> **Replaces the single global delegated walk** with per-root (and per-portal-target) native delivery: each `createRoot` container is an event root, handlers run only on that listener’s logical segment, and native listeners between nested roots interleave and can block the next root. Shadow/slot paths use `composedPath()`; portals still bubble through logical ancestors via portal range bookkeeping and late ownership resolution when children mount hosts after the portal render.
> 
> **Propagation cancellation** now mirrors React’s split between the framework queue and the browser event: handler `stopPropagation()` stops remaining logical handlers and calls native stop; an earlier same-root native stop does not truncate Octane’s queue; native `stopImmediatePropagation()` does not cut off an in-flight logical phase (both stops are needed to halt both). Temporary `stopPropagation`/`currentTarget` wiring is restored after dispatch, including redispatch of the same `Event`.
> 
> **JSX** allows `onCancel` / `onClose` (and capture variants) on non-`dialog` ancestors. **Benchmarks** add a portal mount/unmount gate before the 128-input stress run and assert no leftover portal DOM walks. **Tests** add Chromium React-oracle boundary cases plus conformance coverage for nested roots, deferred portal content, shared targets, and cancellation edge cases. **Docs** and the changeset describe the new semantics and the `stopImmediatePropagation` migration note.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c3c476eb0199301f210cc709fd76bc0e0aba5c4f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/octane/src/runtime.ts [hunk 1]
```diff
@@ -16297,11 +16297,88 @@ const _delegated = new Set<string>();
 // portals are currently rendering into it so we detach only when the last
 // one unmounts. createRoot containers have refcount 1 for their lifetime.
 const _delegationTargets = new Map<Node, number>();
+// Portal descendants can add their first host during a later local update,
```
[the remark this answers is not in the document]
Comment (domgan-oai) [reply packages/octane/src/runtime.ts:16301]: Fixed in c3c476eb0. The compiled portalSlotSlot unmount path now unregisters its portal event range before the existing delegation-target cleanup, without changing detach or transaction behavior.

Added a production deterministic regression covering six compiled portal owners across three shared-target cycles. The e4a0d462 baseline fails with 256 leftover sibling reads after unmount; the fixed runtime has zero. Portal delivery, ref cleanup, detached-node behavior, and all 128 subsequent controlled input/output updates remain correct.

The final source also passes 15,896 core/binding tests, 24 Chromium tests, scoped types, and the paired event work/timing gates. Current-head CI has been restarted.
